source/game/field/RKGndCol.cc

```diff
@@ -0,0 +1,374 @@
+
+        m_bbox.mMin = m_bbox.mMin.minimize(vtx1);
+        m_bbox.mMin = m_bbox.mMin.minimize(vtx2);
+        m_bbox.mMin = m_bbox.mMin.minimize(vtx3);
+        m_bbox.mMax = m_bbox.mMax.maximize(vtx1);
+        m_bbox.mMax = m_bbox.mMax.maximize(vtx2);
+        m_bbox.mMax = m_bbox.mMax.maximize(vtx3);
+    }
+}
+
+EGG::Vector3f RKGndCol::GetVertex(f32 height, const EGG::Vector3f &vertex1,
+        const EGG::Vector3f &fnrm, const EGG::Vector3f &enrm3, const EGG::Vector3f &enrm) {
+    EGG::Vector3f cross = fnrm.cross(enrm);
+    f32 dp = cross.ps_dot(enrm3);
+    cross *= (height / dp);
+
+    return cross + vertex1;
+}
+
+void RKGndCol::lookupSphere(f32 radius, const EGG::Vector3f &pos, const EGG::Vector3f &prevPos,
+        KCLTypeMask typeMask) {
+    f32 radiusClamped = radius;
+    if (radius > this->m_sphereRadius) {
+        radiusClamped = this->m_sphereRadius;
+    }
+    this->m_prismIter = searchBlock(pos);
+    this->m_pos = pos;
+    this->m_prevPos = prevPos;
+    this->m_movement = pos - prevPos;
+    this->m_radius = radiusClamped;
+    this->m_typeMask = typeMask;
+}
+
+const u16 *RKGndCol::searchBlock(const EGG::Vector3f &point) {
+    // Calculate the x, y, and z offsets of the point from the minimum
+    // corner of the tree's bounding box.
+    const int x = point.x - this->m_areaMinPos.x;
+    const int y = point.y - this->m_areaMinPos.y;
+    const int z = point.z - this->m_areaMinPos.z;
+
+    // Check if the point is outside the tree's bounding box in the x, y,
+    // or z dimensions. If it is, return 0.
+    if (x & this->m_areaXWidthMask || y & this->m_areaYWidthMask || z & this->m_areaZWidthMask) {
+        return nullptr;
+    }
+
+    // Initialize the current tree node to the root node of the tree.
+    u32 shift = this->m_blockWidthShift;
+    const u8 *curBlock = reinterpret_cast<const u8 *>(this->m_blockData);
+    s32 offset;
+
+    // Traverse the tree to find the leaf node containing the input point.
+    u32 index = 4 *
+            (((u32)z >> shift) << this->m_areaXYBlocksShift |
+                    ((u32)y >> shift) << this->m_areaXBlocksShift | (u32)x >> shift);
+
+    while (true) {
+        // Get the offset of the current node's child node.
+        offset = parse<u32>(*reinterpret_cast<const u32 *>(curBlock + index));
+
+        // If the offset is negative, the current node is a leaf node.
+        if ((offset & 0x80000000) != 0) {
+            break;
+        }
+
+        // If the offset is non-negative, update the current node to be
+        // the child node and continue traversing the tree.
+        shift--;
+        curBlock += offset;
+
+        index = 4 *
+                (((4 * ((u32)z >> shift)) & 4) | ((2 * ((u32)y >> shift)) & 2) |
+                        ((1 * ((u32)x >> shift)) & 1));
+    }
+
+    // Return the address of the leaf node containing the input point.
+    return reinterpret_cast<const u16 *>(curBlock + (offset & 0x7FFFFFFF));
```

Comment:
Is the 4 byte restriction necessary here? `offset` is parsed as a `u32`, so it's not like it will exceed that.

Reply:
Remember that a leaf node is determined by its leftmost bit. If we include the leftmost bit in the pointer arithmetic, we instantly segfault.

We could likely better communicate this by using an inverse mask.